cns/requestcontroller/kubernetes/k8srequestcontroller.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	//Add CRD scheme to runtime sheme so manager can recognize it
+	var scheme = runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = nnc.AddToScheme(scheme)
+
+	// GetConfig precedence
+	// * --kubeconfig flag pointing at a file at this cmd line
+	// * KUBECONFIG environment variable pointing at a file
+	// * In-cluster config if running in cluster
+	// * $HOME/.kube/config if exists
+	// We're not using GetConfigOrDie becuase we want to propogate the error if there is one
+	k8sconfig, err := ctrl.GetConfig()
+	if err != nil {
+		logger.Errorf("[cns-rc] Error getting kubeconfig: %v", err)
+		return nil, err
+	}
+
+	// Create manager for NodeNetworkConfigReconciler
+	// MetricsBindAddress is the tcp address that the controller should bind to
+	// for serving prometheus metrics, set to "0" to disable
+	mgr, err := ctrl.NewManager(k8sconfig, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		Namespace:          k8sNamespace,
+	})
+	if err != nil {
+		logger.Errorf("[cns-rc] Error creating new request controller manager: %v", err)
+		return nil, err
+	}
+
+	// Create the requestController struct
+	k8sRequestController := k8sRequestController{
+		mgr:         mgr,
+		Client:      mgr.GetClient(),
+		restService: restService,
+		hostName:    hostName,
+	}
+
+	return &k8sRequestController, nil
+}
+
+// StartRequestController starts the reconcile loop. This loop waits for changes to CRD statuses.
+// When a CRD status change is made, Reconcile from nodenetworkconfigreconciler is called.
+func (k8sRC *k8sRequestController) StartRequestController() error {
+	nodenetworkconfigreconciler := &NodeNetworkConfigReconciler{
+		K8sClient:   k8sRC.Client,
+		RestService: k8sRC.restService,
+		HostName:    k8sRC.hostName,
+	}
+
+	// Setup manager with NodeNetworkConfigReconciler
+	if err := nodenetworkconfigreconciler.SetupWithManager(k8sRC.mgr); err != nil {
+		logger.Errorf("[cns-rc] Error creating new NodeNetworkConfigReconciler: %v", err)
+		return err
+	}
+
+	// Start manager and consequently, the reconciler
+	// Start() blocks until SIGINT or SIGTERM is received
+	go func() {
+		logger.Printf("Starting manager")
+		if err := k8sRC.mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+			logger.Errorf("[cns-rc] Error starting manager: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// ReleaseIPsByUUIDs sends release ip request to the API server. Provide the UUIDs of the IP allocations
+func (k8sRC *k8sRequestController) ReleaseIPsByUUIDs(listOfIPUUIDS []string) error {
+	nodeNetworkConfig, err := k8sRC.getNodeNetConfig(k8sRC.hostName, k8sNamespace)
+	if err != nil {
+		logger.Errorf("[cns-rc] Error getting CRD when releasing IPs by uuid %V", err)
+		return err
+	}
+
+	//Update the CRD IpsNotInUse
+	nodeNetworkConfig.Spec.IPsNotInUse = append(nodeNetworkConfig.Spec.IPsNotInUse, listOfIPUUIDS...)
+
+	//Send update to API server
+	if err := k8sRC.updateNodeNetConfig(nodeNetworkConfig); err != nil {
```

Comment:
Is the CNS request controller the only one updating the node net config?

Reply:
Yes